Foreign field operations

mitschabaude opened the pull request:
~~mina branch: `feature/foreign-field-develop` (currently no changes but depends on an open PR)~~
~~bindings: https://github.com/o1-labs/snarkyjs-bindings/pull/41~~

closes #961 
closes #963 

This PR adds a `ForeignField` class, which is created by the `createForeignField` class factory (taking the field modulus as parameter).

`ForeignField` aims for a DX similar to what developers are used to from `Field`, with a subset of methods that should be powerful enough for most use cases. ~~So far, all methods are "safe" and fully constrain their outputs, which allows to use the class with confidence and no inside knowledge. An `Unsafe` API which allows advanced developers to save some constraints will be added in a separate PR.~~

EDIT: More info about recent changes in https://github.com/o1-labs/o1js/pull/985#issuecomment-1833498392

querolita commented on src/lib/foreign-field.ts:
```diff
@@ -0,0 +1,496 @@
+     */
+    toBits(length = sizeInBits) {
+      checkBitLength('ForeignField.toBits()', length, sizeInBits);
+      let [l0, l1, l2] = this.toFields();
```
Just for clarity, `l0` is the least significant limb?

mitschabaude replied:
yes!

MartinMinkov commented on src/lib/foreign-field.ts:
```diff
@@ -0,0 +1,496 @@
+     *
+     * See {@link FieldVar} to understand constants vs variables.
+     *
+     * **Warning**: This function is only useful in `Provable.witness()` or `Provable.asProver()` blocks,
```
Nit: Is it possible to add a link to `Provable.witness` or `Provable.asProver` in the doc comments here?

querolita commented on src/lib/foreign-field.ts:
```diff
@@ -0,0 +1,496 @@
+      let l2 = Field.fromBits(bits.slice(2 * limbSize, 3 * limbSize));
+      let x = ForeignField.fromFields([l0, l1, l2]);
+      // TODO: can this be made more efficient? since the limbs are already range-checked
+      if (length === sizeInBits) x.assertValidElement();
```
In particular, it could be the case that the output is 256 bits for example, but the actual value is greater than the foreign modulus (of secp256k1 for example)

querolita commented on src/lib/foreign-field.ts:
```diff
@@ -0,0 +1,496 @@
+// since p = 2^254 + eps for both Pasta fields with eps small, a fairly tight lower bound is
+// f_max >= sqrt(2^254 * 2^264) = 2^259
+const foreignFieldMaxBits = (BigInt(Fp.sizeInBits - 1) + 3n * limbBits) / 2n;
+const foreignFieldMax = 1n << foreignFieldMaxBits;
```
Actually this is not super super accurate.

We computed with sage the real maximum possible field we can support with pallas and vesta which is the one with prime order `926336713898529563388567880069503262826888842373627227613104999999999999999607 `

MartinMinkov commented on src/lib/foreign-field.ts:
```diff
@@ -0,0 +1,496 @@
+    /**
+     * Checks whether this field element is a constant.
+     *
+     * See {@link FieldVar} tp understand constants vs variables.
```
```suggestion
     * See {@link FieldVar} to understand constants vs variables.
```

MartinMinkov commented on src/lib/foreign-field.ts:
```diff
@@ -0,0 +1,496 @@
+     *
+     * See {@link FieldVar} to understand constants vs variables.
+     *
+     * **Warning**: This function is only useful in `Provable.witness()` or `Provable.asProver()` blocks,
```
```suggestion
     * **Warning**: This function is only useful in {@link Provable.witness}` or {@link Provable.asProver}` blocks,
```

mitschabaude replied:
```suggestion
     * **Warning**: This function is only useful in {@link Provable.witness} or {@link Provable.asProver} blocks,
```

querolita reviewed (commented):
Very well done! I would accept, but let's have @jspada take a look as well.

MartinMinkov commented on src/lib/foreign-field.ts:
````diff
@@ -0,0 +1,496 @@
+
+    /**
+     * Finite field addition
+     *
+     * ```ts
````
```suggestion
     * @example
     * ```ts
```

MartinMinkov commented on src/lib/foreign-field.ts:
````diff
@@ -0,0 +1,496 @@
+
+    /**
+     * Finite field negation
+     *
+     * ```ts
````
```suggestion
     * @example
     * ```ts
```

MartinMinkov commented on src/lib/foreign-field.ts:
````diff
@@ -0,0 +1,496 @@
+
+    /**
+     * Finite field subtraction
+     *
+     * ```ts
````
```suggestion
     * @example
     * ```ts
```

MartinMinkov commented on src/lib/foreign-field.ts:
````diff
@@ -0,0 +1,496 @@
+
+    /**
+     * Finite field multiplication
+     *
+     * ```ts
````
```suggestion
     * @example
     * ```ts
```

MartinMinkov commented on src/lib/foreign-field.ts:
````diff
@@ -0,0 +1,496 @@
+
+    /**
+     * Multiplicative inverse in the finite field
+     *
+     * ```ts
````
```suggestion
     * @example
     * ```ts
```

MartinMinkov commented on src/lib/foreign-field.ts:
````diff
@@ -0,0 +1,496 @@
+
+    /**
+     * Assert equality with a ForeignField-like value
+     *
+     * ```ts
````
```suggestion
     * @example
     * ```ts
```

MartinMinkov commented on src/lib/foreign-field.ts:
````diff
@@ -0,0 +1,496 @@
+
+    /**
+     * Check equality with a ForeignField-like value
+     *
+     * ```ts
````
```suggestion
     * @example
     * ```ts
```

MartinMinkov commented on src/lib/foreign-field.ts:
```diff
@@ -0,0 +1,496 @@
+    // bit packing
+
+    /**
+     * Unpack a field element to its bits, as a `Bool[]` array.
```
```suggestion
     * Unpack a field element to its bits, as a {@link Bool}[] array.
```

MartinMinkov commented on src/lib/foreign-field.ts:
```diff
@@ -0,0 +1,496 @@
+     * This will check that the field element is in the range [0, p),
+     * where p is the foreign field modulus.
+     *
+     * Exception: If {@link createForeignField} is called with `{ unsafe: true }`,
```
```suggestion
     * **Exception**: If {@link createForeignField} is called with `{ unsafe: true }`,
```

MartinMinkov commented on src/lib/foreign-field.ts:
````diff
@@ -0,0 +1,496 @@
+
+    /**
+     * Create a new {@link ForeignField} from a bigint, number, string or another ForeignField.
+     *
+     * ```ts
````
```suggestion
     * @example
     * ```ts
```

MartinMinkov commented on CHANGELOG.md:
```diff
@@ -17,7 +17,9 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
-> no unreleased changes yet
+### Added
+
+- **Foreign field arithmetic** exposed through the `createForeignField()` class factory https://github.com/o1-labs/snarkyjs/pull/985
```
Nit: Do you think it's helpful to have a more descriptive entry in the changelog for such an awesome feature? :D 

Something like:
`
Introduced 'ForeignField', a class for prime order finite fields distinct from the native Field. Supports arithmetic operations, helper methods, and allows optional bypass of automatic validity checks for optimized code. Supports prime moduli up to 259 bits. https://github.com/o1-labs/snarkyjs/pull/985
`

jspada replied:
I think you want to be more restrictive than just checking less than 259 bits, but also check it's less than the max modulus specifically, which is commented as `926336713898529563388567880069503262826888842373627227613104999999999999999607` here: https://github.com/MinaProtocol/mina/blob/cf5664de9cdd8459edc87cc4c9759b3cbe534cf3/src/lib/crypto/kimchi_backend/gadgets/foreign_field.ml#L42

MartinMinkov reviewed (approved):
I am approving on my end! Great work as always :D

mitschabaude reviewed (commented):
forgot to send these comments 😓

mitschabaude commented on src/lib/foreign-field.ts:
```diff
@@ -0,0 +1,496 @@
+ * `createForeignField(p)` takes the prime modulus `p` of the finite field as input, as a bigint.
+ * We support prime moduli up to a size of 259 bits.
+ *
+ * The returned {@link ForeignField} class supports arithmetic modulo `p` (addition and multiplication),
```
I think `p` is more typical to use for a prime number than `f`, would prefer to stick with it for clarity

mitschabaude commented on src/lib/foreign-field.ts:
```diff
@@ -0,0 +1,496 @@
+  if (p <= 0) {
+    throw Error(`ForeignField: expected modulus to be positive, got ${p}`);
+  }
+  if (p > foreignFieldMax) {
```
Thanks! I actually read the RFC and wanted to use the exact number, but sadly I couldn't compute it exactly with JS floats, and I also didn't want to hard-code a number when everything else is derived from simple constants.
So I looked at the difference between 2^259 and the exact max modulus and figured that they are so close that probably no one will ever use a prime modulus larger than 259 bits.

In fact, this is how `926336713898529563388567880069503262826888842373627227613104999999999999999607` looks in hex:
`80000000000000000000000000000000224698fc094cf91b9908b7c12f823fe77`
you can see how close it is to 2^259 which is `80000...000`

mitschabaude replied:
Also, documenting "259 bits" felt simpler / easier to to use than some random large number

querolita replied:
> you can see how close it is to 2^259 which is 80000...000

Yeah exactly, that's why I said 259.5 bits, because that number is more or less sqrt of 2^519

I am fine with restricting it further, but it is good to know for us that actually we do have some margin in case we want to use it in another scenario where 259.5 bits are needed.

mitschabaude replied:
Cool! The max modulus is not that close to the sqrt of 2^519 though. It's much closer to the sqrt of 2^518, which is 2^259, because the pasta moduli are very close to 2^254 (not 2^255)

querolita replied:
Well, just for curiosity, the exact number is computed performing `sqrt(2^264*pallas)` which is about `sqrt(2^519)`, that's why I say 259.5 bits. Then that modulus is the maximum prime number that still satisfies that.

mitschabaude replied:
`pallas` is about 2^254

mitschabaude commented on src/lib/foreign-field.ts:
```diff
@@ -0,0 +1,496 @@
+     */
+    sub(y: ForeignField | bigint | number) {
+      if (this.isConstant() && isConstant(y)) {
+        let z = mod(this.toBigInt() - toFp(y), p);
```
yeah, bigint subtraction works and can be negative, and z will always be positive. the modulo operation `%` in JS can return a negative number, that's why we use our own `mod()` function which is written so that the result is always in [0, p): https://github.com/o1-labs/snarkyjs-bindings/blob/6b069e49b9c80b6004262333e83c33047a8293eb/crypto/finite_field.ts#L26-L30

mitschabaude commented on src/lib/foreign-field.ts:
````diff
@@ -0,0 +1,496 @@
+     * Check equality with a ForeignField-like value
+     *
+     * ```ts
+     * let isXZero = x.equals(0);
````
yep just an example

mitschabaude commented on src/lib/foreign-field.ts:
```diff
@@ -0,0 +1,496 @@
+
+const limbMax = (1n << limbBits) - 1n;
+
+// the max foreign field modulus is f_max = floor(sqrt(p * 2^t)), where t = 3*limbBits = 264 and p is the native modulus
```
good point!

mitschabaude commented on src/lib/foreign-field.ts:
```diff
@@ -0,0 +1,496 @@
+        let z = mod(this.toBigInt() + toFp(y), p);
+        return new ForeignField(z);
+      }
+      let z = Snarky.foreignField.add(this.value, toVar(y), pMl);
```
Thanks for calling that out! (Actually I think `Foreign_field.add` adds only one zero gate)

Yeah, the idea was to land this safe & simple API as a first step, and figure out a more flexible, advanced API in a future PR, after we had the chance to experiment with implementing an optimized gadget ourselves. I think the advanced API will probably live under an `Unsafe` or `Advanced` namespace and expose all possible options.

Providing the `full` flag in the current API didn't feel right to me: the `~full:false` option is not usable by itself without also knowing how to chain additions and without access to the zero gate / the correct bounds check. Apart from that, it makes the full method more error-prone to use. I think `~full:false` is a much lower-level building block that IMO should be a separate method, instead of being mixed with the full `add()` method.

querolita replied:
Okay! In any case, if you want to provide an easy method to chain, there's the `sum_chain` function which does precisely that. The user basically has to include all of the inputs to be added/subtracted, and a list of operands (+/-), and then it chains it for you, with some final checks at the end.

mitschabaude replied:
yeah I thought about your comments more and decided I want to include a simple to use method which exposes `sum_chain`! will add that today

jspada commented on src/lib/foreign-field.unit-test.ts:
```diff
@@ -0,0 +1,186 @@
+class SmallField extends createForeignField(17n) {}
+let x = new SmallField(16);
+x.assertEquals(-1); // 16 = -1 (mod 17)
+x.mul(x).assertEquals(1); // 16 * 16 = 15 * 17 + 1 = 1 (mod 17)
```
cool

jspada commented on src/lib/foreign-field.unit-test.ts:
```diff
@@ -0,0 +1,186 @@
+
+// invalid example - modulus too large
+
+expect(() => createForeignField(1n << 260n)).toThrow(
```
very good

jspada commented on src/lib/foreign-field.unit-test.ts:
```diff
@@ -0,0 +1,186 @@
+  (x, y) => x === y,
+  Random.scalar
+);
+equivalentVoid2(
```
what's the `Void2` mean?

mitschabaude replied:
that we are testing a function which takes two arguments (`2`) and returns nothing (`Void`)

jspada commented on src/lib/foreign-field.unit-test.ts:
```diff
@@ -0,0 +1,186 @@
+let scalarBigint = Fq.random();
+let pointBigint = G.scale(G.generatorMina, scalarBigint);
+
+// perform a "scalar unshift" in foreign field arithmetic,
```
very cool

jspada commented on src/lib/foreign-field.unit-test.ts:
```diff
@@ -0,0 +1,186 @@
+// using foreign field arithmetic should result in much fewer constraints
+let { rows: rows0 } = Provable.constraintSystem(main0);
+let { rows: rows1 } = Provable.constraintSystem(main1);
+expect(rows0 + 100).toBeLessThan(rows1);
```
nice test!

jspada commented on src/lib/foreign-field.unit-test.ts:
```diff
@@ -0,0 +1,186 @@
+let { rows: rows1 } = Provable.constraintSystem(main1);
+expect(rows0 + 100).toBeLessThan(rows1);
+
+// TODO: add when proving works
```
Is there an issue for doing this already?

mitschabaude replied:
the idea was to not land this PR until proving works and we can enable these tests
this is the Mina PR that I'm waiting on: https://github.com/MinaProtocol/mina/pull/12524

jspada reviewed (approved):
nice tests; added some questions and a suggestion

mitschabaude commented:
Update: Motivated by @querolita's comments I added a new method `ForeignField.sum()` which uses `sum_chain` to chain any number of additions and subtractions efficiently. `add()`, `sub()` and `neg()` are now implemented using `sum()`, so we reduce the OCaml API surface while also exposing a more efficient API which is not `Unsafe`

querolita reviewed (approved):
Great!

bkase commented on src/lib/testing/equivalent.ts:
```diff
@@ -0,0 +1,242 @@
+      });
+    });
+  }
+  function equivalentVoid1(
```
nit: `unit` not `void`, see 11.3.1 of https://www.cs.cmu.edu/~rwh/pfpl/2nded.pdf . C and Java made a mistake

mitschabaude replied:
fixed 🫡

rahulghangas commented:
@mitschabaude any update on this? We're currently blocked by it and if possible, would love to help get this through the finish line

mitschabaude commented:
@rahulghangas the latest on this is that
* we finally have all the dependencies for this in  `o1js/main`
* but, decided to rewrite the foreign field gadgets (and others) in TS before releasing them
the rewrite relevant to this PR will likely start later this week and may take a week at most.

so 2-3 weeks is a realistic ETA until you can use this

mitschabaude commented:
This PR is officially revived! For the reviewer's convenience, here is the diff from the last review: https://github.com/o1-labs/o1js/compare/d462e7c...4330a98

I made a bunch of changes to the original `ForeignField` class, the most impactful of what are:
* the bulk of the class was moved _outside_ the class factory function `createForeignField()`. The reason for that is that TS is not good at all about understanding the types of dynamically created classes (most methods end up having `any` inputs in the compiled `d.ts` file, even as they are fully typed  in the source code). By moving the class outside we can use the precise types specified by it. The base class is of course not functional (it doesn't know its field modulus yet). So the pattern I use is to "inject" the dynamic parts into static properties that are possibly undefined, and then lift those to always defined properties by wrapping them with getters that throw an error.
* second big change: instead of just having one class, I introduce three different variants of every foreign field class: `Unreduced`, `AlmostReduced` and `Canonical`. that way, I can have fine-grained control over which method expects which kind of range check on its inputs, and what range check is proved about its return value. to understand this detail, I recommend to read the doccomment on `createForeignField()`, and also to read through the new example: `src/examples/crypto/foreign-field.ts`

Trivo25 reviewed (approved):
nice!

mitschabaude commented on src/lib/foreign-field.ts:
````diff
@@ -0,0 +1,683 @@
+ * const SmallField = createForeignField(17n); // the finite field F_17
+ * ```
+ *
+ * `createForeignField(p)` takes {@link AlmostForeignField} the prime modulus `p` of the finite field as input, as a bigint.
````
huh, this might have been a misplaced `Ctrl+V`
